qiskit/extensions/hamiltonian_gate.py

```diff
@@ -116,11 +116,11 @@ def qasm(self):
         raise ExtensionError("HamiltonianGate has no QASM definition.")
 
     def validate_parameter(self, parameter):
-        """Hamiltonian parameter has to be an ndarray, operator or float."""
-        if isinstance(parameter, (float, int, numpy.ndarray)):
+        """Hamiltonian parameter has to be an ndarray, operator, complex or float."""
+        if isinstance(parameter, (complex, float, int, numpy.ndarray)):
```

Comment:
Hi. I see your point. However, changing `float` to `complex` in `primitive_op.py` only leads to the failing of an existing unit test called `test_matrix_op_parameterized_evolution` in `test_evolution.py`. @Cryoris suggested a try clause with `float` and using `complex` as a catch for that clause.

Reply:
I agree that your fix will certainly make the tests pass, but my worry is that the code is being too lax now.  In this case, passing a value with a non-zero imaginary component as the time parameter _should_ raise an error - it can't be correct, since it would make the operator non-unitary.

Reply:
What's your suggestion for a better approach?

Reply:
My change in the primitive_op could look like this:
`param_value = float(param_value) if param_value.is_real() else complex(param_value)` and hamiltonian_gate would be rolled back to the original version.
primitive_op does not always lead to evolution. So if a complex param gets to hamiltonian_gate, it will be detected as an error and for other uses case, complex might fly.
The problem is that `is_real()` does not really work for bound parameters. For example, `ParameterExpression(5.0*I).is_real()` evaluates to `True`.

Reply:
Yeah, I agree - that's a better way of doing it.